File: algorithms/leetcode/369-plus-one-linked-list/PlusOneLInkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode plusOne(ListNode head) {
        ListNode prev = null;
        ListNode node = head;
        
        while (node != null) {
            ListNode tmp = node.next;
            node.next = prev;
            prev = node;
            node = tmp;
        }
        
        head = null;
        node = prev;
        int carry = 1;
        while (node != null) {
            node.val += carry;
            if (node.val == 10) {
                carry = 1;
                node.val = 0;
            } else {
                carry = 0;
            }
            ListNode tmp = node.next;
            node.next = head;
            head = node;
            node = tmp;
        }
        
        if (carry == 1) {
            ListNode oneHead = new ListNode(carry);
            oneHead.next = head;
            head = oneHead;
        }
        
        return head;
    }
}

File: algorithms/leetcode/369-plus-one-linked-list/plus-one-linked-list.js
const plusOne = head => {
    if (!head) {
        return null;
    }

    let prev = null;

    while(head) {
        const tmp = head.next;
        head.next = prev;
        prev = head;
        head = tmp;
    }

    let leftover = 1;

    head = null;

    while(prev) {
        const tmp = prev.next;
        prev.next = head;

        prev.val += leftover;

        if (prev.val >= 10) {
            leftover = 1;
            prev.val -= 10;
        } else {
            leftover = 0;
        }

        head = prev;
        prev = tmp;
    }

    if (leftover) {
        const tmp = new ListNode(leftover);
        tmp.next = head;
        head = tmp;
    }

    return head;
};
